### PR TITLE
[#6431] changing "Simple Example" to use implode/explode

### PR DESCRIPTION
**Pull request by mccullagh**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | no |
| Applies to | all |
| Fixed tickets | #6431 |

### CODE REVIEW

**Comment by mccullagh** [comment]
I preferred to use `labels`and not `tags`.

File: cookbook/form/data_transformers.rst [hunk 1]
```diff
@@ -16,24 +16,24 @@ to render the form, and then back into a ``DateTime`` object on submit.
     When a form field has the ``inherit_data`` option set, Data Transformers
     won't be applied to that field.
 
-Simple Example: Sanitizing HTML on User Input
+Simple Example: transforming labels string from User Input to array
```
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: Should be `Transforming Labels String from User Input to Array`

File: cookbook/form/data_transformers.rst [hunk 2]
```diff
@@ -63,27 +57,24 @@ class::
...
                     },
-                    function ($submittedDescription) {
-                        // remove most HTML tags (but not br,p)
-                        $cleaned = strip_tags($submittedDescription, '<br><br/><p>');
-
-                        // transform any \n to real <br/>
-                        return str_replace("\n", '<br/>', $cleaned);
+                    function ($submittedLabels) {
+                        // transform the string back to Array
+                        return explode(',', $submittedLabels);
```
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: missing a space after the comma

File: cookbook/form/data_transformers.rst [hunk 3]
```diff
@@ -16,24 +16,24 @@ to render the form, and then back into a ``DateTime`` object on submit.
...
-    use Symfony\Component\Form\Extension\Core\Type\TextareaType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
 
     // ...
     class TaskType extends AbstractType
     {
         public function buildForm(FormBuilderInterface $builder, array $options)
         {
-            $builder->add('description', TextareaType::class);
+            $builder->add('labels', TextType::class);
```
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: Shouldn't this go on 2.3?

File: cookbook/form/data_transformers.rst [hunk 4]
```diff
@@ -46,15 +46,9 @@ Suppose you have a Task form with a description ``textarea`` type::
         // ...
     }
 
-But, there are two complications:
-
-#. Your users are allowed to use *some* HTML tags, but not others: you need a way
-   to call :phpfunction:`strip_tags` after the form is submitted;
+Internally we want to handle the ``labels`` as array, but to have the form simple we wanna allow the User to edit them as a string.
```
Comment (javiereguiluz) [inline cookbook/form/data_transformers.rst]: We don't use "we" in the docs. What if we reword this as:

```
Internally the labels are stored as an array, but they are displayed
to the user as a simple string, to make them easier to edit.
```

File: cookbook/form/data_transformers.rst [hunk 5]
```diff
@@ -63,27 +57,24 @@ class::
...
 
-            $builder->get('description')
+            $builder->get('labels')
                 ->addModelTransformer(new CallbackTransformer(
-                    // transform <br/> to \n so the textarea reads easier
-                    function ($originalDescription) {
-                        return preg_replace('#<br\s*/?>#i', "\n", $originalDescription);
+                    // transform array to string so the input reads easier
+                    function ($originalLabels) {
+                        return implode(',', $originalLabels);
```
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: missing space after the comma `implode(', ', ...`

**Comment by HeahDude** [comment]
`tags` field is used elsewhere in the docs, we should keep it for consistency. What's the advantage of using `labels`?

File: cookbook/form/data_transformers.rst [hunk 6]
```diff
@@ -63,27 +57,24 @@ class::
...
+            $builder->add('labels', TextType::class);
 
-            $builder->get('description')
+            $builder->get('labels')
                 ->addModelTransformer(new CallbackTransformer(
-                    // transform <br/> to \n so the textarea reads easier
-                    function ($originalDescription) {
-                        return preg_replace('#<br\s*/?>#i', "\n", $originalDescription);
+                    // transform array to string so the input reads easier
+                    function ($originalLabels) {
```
Comment (javiereguiluz) [inline cookbook/form/data_transformers.rst]: What if we rename these variables:

```
$originalLabels  -> $labelsAsArray
$submittedLabels -> $labelsAsString
```

This is just a proposal. Please don't make any change until our doc managers approve/reject this idea. Thanks!

Comment (WouterJ) [inline cookbook/form/data_transformers.rst]: I like your proposal

Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: original/submitted keeps an "event" context, and I like that.

**Comment by WouterJ** [comment]
Looks great! I don't care much about labels vs tags. After you applied the open comments, I think it's ready to be merged. Thanks for working on this!

File: cookbook/form/data_transformers.rst (hunk 6 above)
Comment (mccullagh) [inline cookbook/form/data_transformers.rst]: original/submitted helps to identify the source of the Callbacks.

**Comment by javiereguiluz** [comment]
@WouterJ I think consistency is important here. As @HeahDude said, we use `tags` everywhere ... so I'm 👍 to replace labels by tags.

File: cookbook/form/data_transformers.rst [hunk 7]
```diff
@@ -106,10 +97,10 @@ in your code.
 You can also add the transformer, right when adding the field by changing the format
 slightly::
 
-    use Symfony\Component\Form\Extension\Core\Type\TextareaType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
 
     $builder->add(
-        $builder->create('description', TextareaType::class)
+        $builder->create('description', TextType::class)
```
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: Since text type is the default, I would suggest to remove the second argument everywhere to easy the merge in all branches.

Why not also add a tip to handle `null` as an empty string, we could then remove the hint in master and make a PR to document it in `empty_data` option for the text type (ref symfony/symfony#18357)?

Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: I've pushed e3894be, this should really go on 2.3 imo.

File: cookbook/form/data_transformers.rst [hunk 8]
```diff
@@ -16,24 +16,24 @@ to render the form, and then back into a ``DateTime`` object on submit.
...
-    use Symfony\Component\Form\Extension\Core\Type\TextareaType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
 
     // ...
     class TaskType extends AbstractType
     {
         public function buildForm(FormBuilderInterface $builder, array $options)
         {
-            $builder->add('description', TextareaType::class);
+            $builder->add('tags', TextType::class);
```
Comment (xabbuh) [inline cookbook/form/data_transformers.rst]: I agree that we should merge this into the `2.3` branch first. You don't need to create a new PR for this (we can rebase when merging), but can you please change the type here to be simply `text` (we will then update the type when merging things up)?

File: cookbook/form/data_transformers.rst (hunk 7 above)
Comment (mccullagh) [inline cookbook/form/data_transformers.rst]: removing the `TextType`means i need to remove the `data_class`in the example too because else it would not be clear anymore which Type it is.

File: cookbook/form/data_transformers.rst [hunk 9]
```diff
@@ -16,24 +16,24 @@ to render the form, and then back into a ``DateTime`` object on submit.
     When a form field has the ``inherit_data`` option set, Data Transformers
     won't be applied to that field.
 
-Simple Example: Sanitizing HTML on User Input
----------------------------------------------
+Simple Example: Transforming tags string from User Input to array
```
Comment (xabbuh) [inline cookbook/form/data_transformers.rst]: According to our headline standards this should be "Simple Example: Transforming String Tags from User Input to an Array".

File: cookbook/form/data_transformers.rst [hunk 10]
```diff
@@ -16,24 +16,24 @@ to render the form, and then back into a ``DateTime`` object on submit.
     When a form field has the ``inherit_data`` option set, Data Transformers
     won't be applied to that field.
 
-Simple Example: Sanitizing HTML on User Input
----------------------------------------------
```
Comment (xabbuh) [inline cookbook/form/data_transformers.rst]: Can you please add a label to be BC with the old headline? This should look like this:

```rst
.. _simple-example-sanitizing-html-on-user-input:
```

File: cookbook/form/data_transformers.rst (hunk 7 above)
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: You can still pass `null`

**Comment by WouterJ** [comment]
:+1: 

If you don't have much time, we can fix @xabbuh's comments during the merge as well.

**Comment by mccullagh** [comment]
I have not added a label to be BC with the old headline. I'm not sure how it should be in reStructuredText.

File: cookbook/form/data_transformers.rst [hunk 11]
```diff
@@ -63,27 +57,23 @@ class::
...
+            $builder->add('tags', 'text');
 
-            $builder->get('description')
+            $builder->get('tags')
                 ->addModelTransformer(new CallbackTransformer(
-                    // transform <br/> to \n so the textarea reads easier
-                    function ($originalDescription) {
-                        return preg_replace('#<br\s*/?>#i', "\n", $originalDescription);
+                    // transform array to string so the input reads easier
+                      function ($originalTags) {
```
Comment (HeahDude) [inline cookbook/form/data_transformers.rst]: You should revert the extra spaces here

**Comment by WouterJ** [comment]
@mccullagh you do that by adding the line written by @xabbuh before the new headline. So you end up with this:

```rst
.. _simple-example-sanitizing-html-on-user-input:

Simple Example: Transforming String Tags from User Input to an Array
-------------------------------------------------------------------
```

**Comment by WouterJ** [comment]
Looks perfect, :+1: !

**Comment by xabbuh** [comment]
👍

**Comment by WouterJ** [comment]
Thanks @mccullagh for writing this! I've merged this into the 2.3 branch via d042351 and made some minor improvements afterwards in c8f4583 . I'll now merge them into the newer branches and update the form type names.

**Comment by mccullagh** [comment]
Thank you @WouterJ and everyone who helped.
